test/metabase/lib/swap_test.cljc

```diff
@@ -0,0 +1,123 @@
+(ns metabase.lib.swap-test
+  (:require
+   [clojure.test :refer [deftest is testing]]
+   [clojure.test.check.clojure-test :refer [defspec]]
+   [clojure.test.check.generators :as gen]
+   [clojure.test.check.properties :as prop]
+   [medley.core :as m]
+   [metabase.lib.core :as lib]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.test-metadata :as meta]
+   [metabase.test.util.log :as tu.log]
+   #?@(:cljs ([metabase.test-runner.assert-exprs.approximately-equal]))))
+
+#?(:cljs (comment metabase.test-runner.assert-exprs.approximately-equal/keep-me))
+
+(defn- swapped-clauses-prop [clause-fn clause-key query]
+  (prop/for-all*
+    [(gen/elements (clause-fn query))
+     (gen/elements (clause-fn query))]
+    (fn [source-clause target-clause]
+      (or (= source-clause target-clause) ; Skip any case where we happened to draw the same one twice.
+          (let [swapped (lib/swap-clauses query -1 source-clause target-clause)]
+            (and ;; Correctly rearranged the clauses.
+                 (= (for [clause (clause-fn query)]
+                      (cond
+                        (= source-clause clause) target-clause
+                        (= target-clause clause) source-clause
+                        :else                    clause))
+                    (clause-fn swapped))
+                 ;; And didn't change anything else.
+                 (= (m/dissoc-in query   [:stages 0 clause-key])
+                    (m/dissoc-in swapped [:stages 0 clause-key]))))))))
+
+(defspec swap-clauses-on-aggregations-test-permutations
+  (swapped-clauses-prop
+    lib/aggregations :aggregation
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/aggregate (lib/count))
+        (lib/aggregate (lib/sum (meta/field-metadata :orders :subtotal)))
+        (lib/aggregate (lib/avg (meta/field-metadata :orders :discount)))
+        (lib/aggregate (lib/max (meta/field-metadata :orders :tax)))
+        (lib/aggregate (lib/min (meta/field-metadata :orders :total))))))
+
+(defspec swap-clauses-on-breakouts-test-permutations
+  (swapped-clauses-prop
+    lib/breakouts :breakout
+    (as-> (lib/query meta/metadata-provider (meta/table-metadata :orders)) $q
+      (lib/breakout $q (meta/field-metadata :products :category))
+      (lib/breakout $q (meta/field-metadata :people :source))
+      (lib/breakout $q (lib/with-binning
```

Comment:
Would be nice to test the same field with different/no levels of binning

Reply:
Done.